protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/unmarshall/AwsQueryErrorProtocolUnmarshaller.java

```diff
@@ -0,0 +1,301 @@
+    }
+
+    /**
+     * @return Builder for the default service exception. Used when the error code doesn't match
+     * any known modeled exception or when we can't determine the error code.
+     */
+    private AwsServiceException.Builder defaultException() {
+        return (AwsServiceException.Builder) defaultExceptionSupplier.get();
+    }
+
+    /**
+     * Unmarshalls the XML into the appropriate modeled exception based on the error code. If the error code
+     * is not present or does not match any known exception we unmarshall into the base service exception.
+     *
+     * @param errorRoot Root of <Error/> element. Contains any modeled fields of the exception.
+     * @param errorCode Error code identifying the modeled exception.
+     * @return Unmarshalled exception builder.
+     */
+    private AwsServiceException.Builder unmarshallFromErrorCode(SdkHttpFullResponse response,
+                                                                XmlElement errorRoot,
+                                                                String errorCode) {
+        SdkPojo sdkPojo = exceptions.getOrDefault(errorCode, defaultExceptionSupplier).get();
+        AwsServiceException.Builder builder =
+            ((AwsServiceException) errorUnmarshaller.unmarshall(sdkPojo, errorRoot, response)).toBuilder();
+        builder.message(getMessage(errorRoot));
+        return builder;
+    }
+
+    /**
+     * Extracts the error code (used to identify the modeled exception) from the <Error/>
+     * element.
+     *
+     * @param errorRoot Error element root.
+     * @return Error code or null if not present.
+     */
+    private String getErrorCode(Optional<XmlElement> errorRoot) {
+        return errorRoot.map(e -> e.getOptionalElementByName("Code")
+                                   .map(XmlElement::textContent)
+                                   .orElse(null))
+                        .orElse(null);
+    }
+
+    /**
+     * Extracts the error message from the XML document. The message is in the <Error/>
+     * element for all services.
+     *
+     * @param errorRoot Error element root.
+     * @return Error message or null if not present.
+     */
+    private String getMessage(XmlElement errorRoot) {
+        return errorRoot.getOptionalElementByName("Message")
+                        .map(XmlElement::textContent)
+                        .orElse(null);
+    }
+
+    /**
+     * Extracts the request ID from the XML document. Request ID is a top level element
+     * for all protocols, it may be RequestId or RequestID depending on the service.
+     *
+     * @param document Root XML document.
+     * @return Request ID string or null if not present.
+     */
+    private String getRequestId(SdkHttpFullResponse response, XmlElement document) {
+        XmlElement requestId = document.getOptionalElementByName("RequestId")
+                                       .orElse(document.getElementByName("RequestId"));
```

Comment:
Method in orElse is same as underlying method of getOptionalElementByName? Can we just return null in orElse()?

Reply:
good catch, the second should have been RequestID capital D. EC2 and SimpleDB have a different casing then most XML/Query services.